File: Calculo Numerico/Trabalho 08/Trabalho8.cpp
/*
* Trabalho 8.
* Implemente o Método de Newton.
*
* Resolva o método para um conjunto 
* de pelo menos 5 pares de dados (xi,yi).
*
* Calculo Numerico
* Claudio Schepke
* ---
* Ciencia da Computacao
* Universidade Federal do Pampa
* 
*
* Integrantes:
* ---
* Marcos V. Treviso, 121150107, marcosvtreviso@gmail.com
*
*
* Instrucoes:
* ---
* Compile: g++ -o trab Trabalho8.cpp -Wall -Wextra
* Run: ./trab
*
*/

#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <cstring>
#include <vector>

using namespace std;




const bool DEBUG = false;
const int N = 5;
double x[N] = {0.2, 0.4, 0.6, 0.8, 1.0};
double y[N] = {1.22, 1.49, 1.82, 2.23, 2.72};
double tabela[N][N];

void newton(){

	// caso base
	for(int i=0; i<N; i++)
		tabela[i][0] = y[i];


	for(int i=1; i<N; i++){
		for(int j=i; j<N; j++){
			tabela[j][i] = (tabela[j][i-1]-tabela[j-1][i-1])/(x[j]-x[j-i]);
			if(DEBUG)
				printf("(%.2f - %.2f)/(%.2f - %.2f) = %.2f\n", tabela[j][i-1], tabela[j-1][i-1], x[j], x[j-i], tabela[j][i]);
		}
		if(DEBUG)
			printf("\n");
	}
	

	for(int i=0; i<N; i++){
		for(int j=0; j<N; j++){
			if(j > i)
				printf("  ~  ");
			else
				printf("%.2f ", tabela[i][j]);
		}
		printf("\n");
	}
}


double phi(double val_x){

	double total, prod;	
	total = y[0];
	for(int i=1; i<N; i++){

		prod = 1;

		if(DEBUG)
			printf("(%.2lf)", tabela[i][i]);
	

		for(int j=0; j<i; j++){	
			prod *= (val_x - x[j]);
			
			if(DEBUG)
				printf(".(%.2lf - %.2lf)", val_x, x[j]);
		}

		if(DEBUG)
			printf("\n");

		total += (prod*tabela[i][i]);
	}

	return total;
}


void printVet(double *v){
	for(int i=0; i<N; i++)
		printf("%4.2f ", v[i]);
}



int main(){

	double val_x;


	cout << "Metodo de Newton para valores: " << endl;
	cout << "X: "; printVet(x);
	cout << endl;
	cout << "Y: "; printVet(y);
	cout << endl << endl;

	cout << "Tabela de solucao: " << endl;
	newton();
	cout << endl;

	while(true){
			
		cout << "Informe o valor de X ou CTRL-C para parar:" << endl;
		cin >> val_x;

		cout << "Phi(X): " << endl;
		cout << phi(val_x) << endl << endl;
	}

	return 0;
}

File: Calculo Numerico/Trabalho 08/Trabalho8.py
import math

n = 6
x = [0.15, 0.20, 0.25, 0.30, 0.35, 0.40]
y = [0.12, 0.16, 0.19, 0.22, 0.25, 0.27]
mat = [[0 for y in range(n)] for x in range(n)]

def newton():

	for i in range(n):
		mat[i][0] = y[i]

	for i in range(1, n):
		for j in range(i, n):
			mat[j][i] = (mat[j][i-1]-mat[j-1][i-1])/(x[j]-x[j-i])
	

	for i in range(n):
		for j in range(n):
			if(j > i):
				print("%4.4s " % "~", end="")
			else:
				print("%4.4f " % mat[i][j], end="")
		print("")



def roundp(number, precision):
	return  ("{:."+str(precision)+"f}").format(number) 


def polynomial(p=4):
	
	polynom = roundp(y[0], p)
	for i in range(1, n):
		polynom += " + "
		for j in range(0, i):
			polynom += "(x - "+roundp(x[j], p)+")*"
		polynom += "("+roundp(mat[i][i], p)+")"

	return polynom


def phi(valorX):

	total = y[0]
	for i in range(1, n):
		prod = 1.0
		for j in range(0, i):
			prod *= (valorX - x[j])
		total += (prod*mat[i][i])

	return total


print("Metodo de Newton para valores: ")
print("X: ", end="")
print(x)
print("Y: ", end="")
print(y, end="\n\n")

print("Tabela de solucao: ")
newton();
print("")

print("Polinomio gerador: ")
print("Phi(x) = "+polynomial())
print("")

while(True):
	print("Digite o valor de X ou `CTRL-C` para parar:")

	valorX = float(input())
	print("Resultado: %.4lf\n" % phi(valorX))
